apm-agent-core/src/main/java/co/elastic/apm/agent/bci/bytebuddy/postprocessor/AssignToPostProcessorFactory.java

```diff
@@ -108,33 +112,59 @@ public StackManipulation resolve(TypeDescription typeDescription, MethodDescript
+        return new Advice.PostProcessor() {
+            @Override
+            public StackManipulation resolve(TypeDescription instrumentedType, MethodDescription instrumentedMethod, Assigner assigner, Advice.ArgumentHandler argumentHandler) {
+                if (assignToThrown.index() != -1) {
+                    return objectArrayStackManipulation(adviceMethod, assigner, false, argumentHandler, exit, assignToThrown.index(), TypeDescription.THROWABLE.asGenericType(), MethodVariableAccess.REFERENCE.storeAt(argumentHandler.thrown()));
+                } else {
+                    return scalarStackManipulation(adviceMethod, assigner, false, argumentHandler, exit, TypeDescription.THROWABLE.asGenericType(), MethodVariableAccess.REFERENCE.storeAt(argumentHandler.thrown()), assignToThrown.typing());
+                }
+            }
+        };
+    }
+
+    private StackManipulation objectArrayStackManipulation(MethodDescription.InDefinedShape adviceMethod, Assigner assigner, boolean loadThis, Advice.ArgumentHandler argumentHandler, boolean exit,
+                                                           int index, TypeDescription.Generic targetType, StackManipulation store) {
+        if (!adviceMethod.getReturnType().represents(Object[].class)) {
+            throw new IllegalStateException("Advice method has to return Object[] when setting an index");
+        }
+        StackManipulation load = MethodVariableAccess.REFERENCE.loadFrom(exit ? argumentHandler.exit() : argumentHandler.enter());
+        return new StackManipulation.Compound(
+            loadThis ? MethodVariableAccess.loadThis() : StackManipulation.Trivial.INSTANCE,
+            load,
+            IntegerConstant.forValue(index),
+            ArrayAccess.REFERENCE.load(),
+            assigner.assign(TypeDescription.Generic.OBJECT, targetType, Assigner.Typing.DYNAMIC),
+            store
+        );
+    }
+
+    private StackManipulation scalarStackManipulation(MethodDescription.InDefinedShape adviceMethod, Assigner assigner, boolean loadThis, Advice.ArgumentHandler argumentHandler,
+                                                      boolean exit, TypeDescription.Generic target, StackManipulation store, Assigner.Typing typing) {
+        final StackManipulation assign = assigner.assign(adviceMethod.getReturnType(), target, typing);
+        if (!assign.isValid()) {
+            throw new IllegalStateException("Cannot assign " + adviceMethod.getReturnType() + " to " + target + " in advice method " + adviceMethod.toGenericString());
+        }
+        StackManipulation load = MethodVariableAccess.of(adviceMethod.getReturnType()).loadFrom(exit ? argumentHandler.exit() : argumentHandler.enter());
+        return new StackManipulation.Compound(
+            loadThis ? MethodVariableAccess.loadThis() : StackManipulation.Trivial.INSTANCE,
+            load,
+            assign,
+            store
+        );
+    }
```

Comment:
:exploding_head: I will trust you on this, congrats on refactoring this to make it more understandable.